src/core/mount.c

```diff
@@ -1486,7 +1488,8 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
          * race, let's explicitly scan /proc/self/mountinfo before we start processing /usr/bin/(u)mount
          * dying. It's ugly, but it makes our ordering systematic again, and makes sure we always see
          * /proc/self/mountinfo changes before our mount/umount exits. */
-        (void) mount_process_proc_self_mountinfo(u->manager);
+        if (drain_libmount(u->manager) > 0)
+                (void) mount_process_proc_self_mountinfo(u->manager);
```

Comment:
this should be a separate comit I think. Makes a ton of sense, but should be independent of the rest.